### PR TITLE
[BREAKING CHANGE] webpack 4

### PR DESCRIPTION
**Pull request by Hypnosphi**
Issues: #3044, #3083, #3085, #3092, #3115, #3116, #3135

Things that block this right now

- [x] stable `3.4.0` release (we don't want to maintain 3 versions at the same time)
- [x] ~`react-dev-utils`, namely`InterpolateHtmlPlugin` and `WatchMissingNodeModulesPlugin`. See https://github.com/facebook/create-react-app/pull/4077~ (actually they work as is now)
- [x] #3161
- [x] Fix error in angular-cli example https://github.com/storybooks/storybook/pull/3251

### CODE REVIEW

**Comment by Hypnosphi** [comment]
Weird, storybook build for polymer works OK locally on my machine, but on CI it's failing:

```
ERROR in /tmp/storybook/addons/knobs/dist/polymer/WrapStory.html
Module not found: Error: Can't resolve 'polymer-webpack-loader/register-html-template' in '/tmp/storybook/addons/knobs/dist/polymer'
 @ /tmp/storybook/addons/knobs/dist/polymer/WrapStory.html 7:27-83
 @ /tmp/storybook/addons/knobs/dist/polymer/index.js
 @ /tmp/storybook/addons/knobs/polymer.js
 @ ./src/stories/addon-knobs.stories.js
 @ ./src/stories sync \.stories\.js$
 @ ./.storybook/config.js
 @ multi /tmp/storybook/app/polymer/dist/server/config/polyfills.js /tmp/storybook/app/polymer/dist/server/config/globals.js /tmp/storybook/node_modules/webpack-hot-middleware/client.js?reload=true ./.storybook/config.js
```

**Comment by Hypnosphi** [comment]
For now, I've just disabled `InterpolateHtmlPlugin` and `WatchMissingNodeModulesPlugin` which means that first 4.0 alpha will probably lack the corresponding features: `.env` support in `*-head.html`, and watching for newly installed packages

**Comment by igor-dv** [comment]
Both plugins are rather simple, maybe it worth copypasting/fixing them into the core until the official support?

**Comment by Hypnosphi** [comment]
They say they're almost ready: https://github.com/facebook/create-react-app/pull/4077#issuecomment-370789259

**Comment by pkuczynski** [comment]
Could you also in this work stream upgrade `babel` to `7.0` (currently in `beta.40`) and `babel-loader` to 8.0.0-beta.2`? That allows using more dynamic `.babelrc.js`...

**Comment by Hypnosphi** [comment]
@pkuczynski sure, definitely makes sense. I think it'll be a separate PR though

**Comment by pkuczynski** [comment]
Cool! I can provide a PR even now, as this is not related to webpack and I am being able to successfully use `babel@7` with `webpack@3` in my project...

And why do you need to wait for https://github.com/babel/babel/pull/7472?

**Comment by Hypnosphi** [comment]
Because it will solve a lot of issues we currently have because of custom babelrc resolution

See e.g. #754 #2635 #2610

**Comment by pkuczynski** [comment]
But then it would require quite some work on your side I guess? While basic upgrade to babel@7 should work out of the box now. So maybe it’s worth to do it in two steps?

**Comment by Hypnosphi** [comment]
Probably. Feel free to give it a try

File: app/react/src/server/config/webpack.config.prod.js [hunk 1]
```diff
@@ -47,21 +46,8 @@ export default function(configDir) {
...
-          ie8: false,
-          mangle: false,
-          warnings: false,
-          compress: {
-            keep_fnames: true,
-          },
-          output: {
-            comments: false,
-          },
-        },
```
Comment (Hypnosphi) [inline app/react/src/server/config/webpack.config.prod.js]: @ndelangen should we replicate those settings in `optimization.minimizer` option, or are we OK with webpack [defaults](https://github.com/webpack/webpack/blob/4f399322ad1922ae7bafdeea1c4d83bf9c9851a7/lib/WebpackOptionsDefaulter.js#L271-L276)?

**Comment by Hypnosphi** [comment]
We'll be able to add chunks splitting after either https://github.com/jantimon/html-webpack-plugin/issues/882 or https://github.com/webpack/webpack/issues/6717 resolves

File: app/react/src/server/config/webpack.config.prod.js (hunk 1 above)
Comment (ndelangen) [reply app/react/src/server/config/webpack.config.prod.js]: Let's try to use their defaults!

**Comment by TomFoyster** [comment]
Is there any timeframe for this? We'd love to use WP4 in our production code, but as the component lib uses Storybook we can't progress.

**Comment by Hypnosphi** [comment]
I think it's ready to merge as is, so we could release `4.0.0-alpha` ASAP
I'm also going to release a temporary fork of `react-dev-utils` under `@storybook` scope

@shilman WDYT

**Comment by Cryrivers** [comment]
How could I test this branch?

File: app/vue/package.json [hunk 2]
```diff
@@ -43,15 +43,14 @@
     "find-cache-dir": "^1.0.0",
     "global": "^4.3.2",
     "html-loader": "^0.5.5",
-    "html-webpack-plugin": "^2.30.1",
+    "html-webpack-plugin": "^3.0.7",
     "json5": "^0.5.1",
     "markdown-loader": "^2.0.2",
     "react": "^16.2.0",
-    "react-dev-utils": "^5.0.0",
+    "react-dev-utils": "zanettin/react-dev-utils",
```
Comment (danielduan) [inline app/vue/package.json]: is this a repo we're waiting for changes to be merged and published to npm for `react-dev-utils`

File: app/angular/package.json [hunk 3]
```diff
@@ -41,19 +41,18 @@
...
     "html-loader": "^0.5.5",
-    "html-webpack-plugin": "^2.30.1",
+    "html-webpack-plugin": "^3.0.7",
     "json5": "^0.5.1",
     "markdown-loader": "^2.0.2",
     "node-sass": "^4.8.3",
     "raw-loader": "^0.5.1",
     "react": "^16.0.0",
-    "react-dev-utils": "^5.0.0",
+    "react-dev-utils": "zanettin/react-dev-utils",
```
Comment (ndelangen) [inline app/angular/package.json]: Is this something we could revert back now?

**Review by ndelangen — CHANGES_REQUESTED** [review]

**Comment by ndelangen** [comment]
Besides the forked `react-dev-utils` all looks perfect to me.

If we need to keep the fork, we should publish a forked version, I'd rather not ship with installs from github, that's too unpredictable and users behind a proxy might not be able to install.

File: app/angular/package.json (hunk 3 above)
Comment (Hypnosphi) [reply app/angular/package.json]: https://github.com/facebook/create-react-app/pull/4077 is still open

**Comment by Hypnosphi** [comment]
> we should publish a forked version

OK, I'll do that

**Review by ndelangen — APPROVED** [review]

**Review by danielduan — APPROVED** [review]

**Comment by DevanB** [comment]
Super excited to see this released somewhere. One of our project's Storybook is broken due to Babel@7 and Webpack@4. 

🎉

**Comment by pkuczynski** [comment]
This PR does not include Babel@7...

**Comment by Cryrivers** [comment]
Very excited to see this merged. Could we have a prerelease on npm now?

**Comment by Hypnosphi** [comment]
Cc @shilman

**Comment by TomFoyster** [comment]
What are the chances this'll make it into 3.4.0?

Really keen to see this up on NPM.

**Comment by Cryrivers** [comment]
@TomFoyster I believe it goes into `4.0.0` as it's a breaking change.

**Comment by TomFoyster** [comment]
@Cryrivers You're right, I imagine it will. A shame as we're chomping at the bit to get WP4 into our work, but can't without Storybook's support.

**Comment by pkuczynski** [comment]
And what's the timeline for getting this released? I am blocked with migrating my project to WP4 too :(

**Comment by danielduan** [comment]
@ndelangen @Hypnosphi @shilman let's get an alpha `4.0` published?

If @shilman is unavailable, I think we should have a backup person as well.

**Comment by Hypnosphi** [comment]
Ok, I'll do `3.4.0-rc` and `4.0.0-alpha` releases tomorrow if @shilman doesn't become available

**Comment by Hypnosphi** [comment]
Released as [`v4.0.0-alpha.0`](https://github.com/storybooks/storybook/releases/tag/v4.0.0-alpha.0). Please try it and open an issue if something goes wrong

**Comment by ishantdigitalgenius** [comment]
Tried the latest version with the following versions
```
    "@storybook/addon-actions": "^3.3.0",
    "@storybook/addon-info": "^3.3.0",
    "@storybook/addon-knobs": "^3.3.0",
    "@storybook/cli": "^v4.0.0-alpha.0",
    "@storybook/react": "^v4.0.0-alpha.0",
```

When running storybooks, the module building stops at 95% with the following line.

```
95% emitting unnamed compat plugin
```

**Comment by junish** [comment]
> 95% emitting unnamed compat plugin

I am facing with the same error.

I guess that the error is related to the issue of html-webpack-pugin:
https://github.com/jantimon/html-webpack-plugin/issues/870

**Comment by AndyBan** [comment]
It gives you a bit more info when you kill it:

```
95% emitting unnamed compat pluginTypeError: missingDeps.some is not a function
    at compiler.plugin (C:\Javascript\node_modules\@storybook\react\node_modules\react-dev-utils\WatchMissingNodeModulesPlugin.js:25:23)
    at _err1 (eval at create (C:\Javascript\node_modules\tapable\lib\HookCodeFactory.js:24:12), <anonymous>:26:1)
    at Promise.resolve.then.then.then.then.then.then.then.then.catch.then.then.then (C:\Javascript\node_modules\html-webpack-plugin\index.js:222:11)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:160:7)
```

**Comment by Hypnosphi** [comment]
@ishantdigitalgenius @AndyBan can someone of you please create a GitHub repo with minimal reproduction and open an issue with link to it?

**Comment by SleepWalker** [comment]
@Hypnosphi this happens in `react-dev-utils`:

https://github.com/facebook/create-react-app/blob/next/packages/react-dev-utils/WatchMissingNodeModulesPlugin.js#L20-L31

This thing tries to loop over `missingDeps` of type `SortableSet`, which does not have method `some()`:

https://github.com/webpack/webpack/blob/master/lib/util/SortableSet.js

Due to error being thrown, the `callback` won't be called

**Comment by Hypnosphi** [comment]
Any idea why does it work for our [example apps](https://github.com/storybooks/storybook/tree/master/examples)?

**Comment by SleepWalker** [comment]
The next error will be here:

https://github.com/facebook/create-react-app/blob/next/packages/react-dev-utils/InterpolateHtmlPlugin.js#L37

```
ERROR in   TypeError: callback is not a function
  
  - InterpolateHtmlPlugin.js:37 compilation.plugin
    [fc-app]/[react-dev-utils]/InterpolateHtmlPlugin.js:37:11
```

This happens because `html-webpack-plugin` does not provide `callback` argument anymore:

https://github.com/jantimon/html-webpack-plugin/blob/3afd5d228a2775282e751c56eb81c3bf1a2f7b05/index.js#L177-L178
https://github.com/jantimon/html-webpack-plugin/blob/a7d37ca472cb5d9ee9dd58932b169e936416f479/spec/BasicSpec.js#L920-L931

Looks like this is due to new plugin system of webpack4


I have fixed both errors in a very dumb way:

```js
      if (missingDeps.some && missingDeps.some(file => file.indexOf(nodeModulesPath) !== -1)) {

      callback && callback(null, data);
```

After this durty hack the storybook have successfully started

**Comment by SleepWalker** [comment]
Here is more clever candidates for fix:

`InterpolateHtmlPlugin.js`:

```js
...
  apply(compiler) {
    compiler.plugin('compilation', compilation => {
      const handler = (data, callback) => {
        // Run HTML through a series of user-specified string replacements.
        Object.keys(this.replacements).forEach(key => {
          const value = this.replacements[key];
          data.html = data.html.replace(
            new RegExp('%' + escapeStringRegexp(key) + '%', 'g'),
            value
          );
        });
        callback(null, data);
      };

      if (compilation.hooks) {
        compilation.hooks.htmlWebpackPluginBeforeHtmlProcessing
          .tapAsync('InterpolateHtmlPlugin', handler);
      } else {
        compilation.plugin('html-webpack-plugin-before-html-processing', handler);
      }
    });
  }
```

`WatchMissingNodeModulesPlugin.js`:

I've converted Set into an Array, but I'm not sure whether webpack have changed only underlying type or set item type too. I hope that there are still strings as before :)

```js
  apply(compiler) {
    compiler.plugin('emit', (compilation, callback) => {
      var missingDeps = compilation.missingDependencies;
      var nodeModulesPath = this.nodeModulesPath;

      // If any missing files are expected to appear in node_modules...
      if (Array.from(missingDeps).some(file => file.indexOf(nodeModulesPath) !== -1)) {
        // ...tell webpack to watch node_modules recursively until they appear.
        compilation.contextDependencies.push(nodeModulesPath);
      }

      callback();
    });
  }
```

**Comment by Hypnosphi** [comment]
It's actualy already fixed in https://github.com/facebook/create-react-app/pull/4077 and we've released it from our fork as `@storybook/react-dev-tools`. Unfortunately, we have github path [stored] in our lockfile, which made me think that `5.0.0` works again because of some compatibility fix in webpack.

I'll send a hotfix PR ASAP

**Comment by danielduan** [comment]
@ishantdigitalgenius try to use the same version of addons as the rest of the `@storybook` packages. even though there usually aren't any issues, we don't guarantee different versions working together.

**Comment by Hypnosphi** [comment]
Fix released as [`4.0.0-alpha.1`](https://github.com/storybooks/storybook/releases/tag/v4.0.0-alpha.1)

**Comment by peet86** [comment]
With 4.0.0-alpha.1 I still have the [jantimon/html-webpack-plugin#870](https://github.com/jantimon/html-webpack-plugin/issues/870) issue.. :(

` 95% emitting HtmlWebpackPlugin/[..]/node_modules/@storybook/react/node_modules/toposort/index.js:29
      throw new Error('Cyclic dependency: '+JSON.stringify(node))
                                                 ^
TypeError: Converting circular structure to JSON`

**Comment by Hypnosphi** [comment]
We probably need to apply workaround from https://github.com/jantimon/html-webpack-plugin/issues/870#comment-370004105

For now, you can try to apply it from custom webpack config (full control mode):

```
export (baseConfig, env, defaultConfig) => ({
  ... defaultConfig,
  plugins: defaultConfig.plugins.map(plugin => {
    if (plugin.constructor.name === 'HtmlWebpackPlugin') {
      return { ...plugin, chunksSortMode: 'none' }
    }
    return plugin
  })
})
```

**Comment by andrerpena** [comment]
Thank you! Storybook is now working with Webpack 4 and custom Webpack configs but, just so you know, I had to add `babel-core` as a `dev-dependency` to my project in order for it to work.

**Comment by igor-dv** [comment]
`babel-core` is a peer dep, so you should add it

**Comment by Hypnosphi** [comment]
We strongly recommend to use the same versions for `@storybook/<framework>` and addons

**Comment by VRedondo** [comment]
Following settings doesn't work for me

```
"devDependencies": {
  "webpack": "^4.12.0",
  "@storybook/angular": "^4.0.0-alpha.18",
  "babel-core": "^6.26.3",
  "babel-loader": "^8.0.0",
```

when running storybook it returns the error 

```

ERROR in ./.storybook/config.js
Module build failed (from ./node_modules/babel-loader/lib/index.js):
Error: Cannot find module '@babel/core'
```

**Comment by SleepWalker** [comment]
@VRedondo you need to use `babel@7` to run the latest alpha, or use the version with babel 6. I've used `alpha14`, but the most recent is `alpha16`

**Comment by VRedondo** [comment]
Thanks @SleepWalker and @Hypnosphi actually both were needed, babel-loader 7 requires babel 7 so it got fixed with
"@babel/core": "^7.0.0",
"babel-loader": "^8.0.0",

**Comment by Hypnosphi** [comment]
> babel-loader 7 requires babel 7 

What made you think so?

**Comment by VRedondo** [comment]
@Hypnosphi the console output says 
```
Module build failed (from ./node_modules/babel-loader/lib/index.js):
Error: Requires Babel "^7.0.0-0", but was loaded with "6.26.3".
``` 
when running with babel-loader 7 in case you don't have babel 7

**Comment by Hypnosphi** [comment]
Yeah, you also need your own babelrc in this case, our default config is for Babel 7

**Pull request merged**